.github/scripts/rust/setup.py

```diff
@@ -0,0 +1,151 @@
+from fnmatch import fnmatch
+from pathlib import Path
+import re
+import json
+
+CWD = Path.cwd()
+
+# All jobs for all crates will run if any of these paths changes
+ALWAYS_RUN_PATTERNS = ["**/rust-toolchain.toml", ".github/**"]
+
+# Crates which will be tested in release mode
+BENCH_CRATES = ["packages/engine"]
+
+# Exclude the stable channel for these crates
+DISABLE_STABLE_PATTERNS = ["packages/engine**"]
+
+# Try and publish these crates when their version is changed in Cargo.toml
+PUBLISH_PATTERNS = ["packages/libs/error-stack"]
+
+
+def generate_diffs():
```

Comment:
We encountered a bug the other day where the gating of the job wasn't working properly. Have you had a chance to think about it and see if it affects the solution you've put here?

Reply:
If we end up keeping it as is can you have another go at the documentation of this function, just to explain the merge commit case in a bit more detail?

Reply:
As discussed off-Github, there's some confusion coming about here because this function has branching logic to try and make it useable for local changes, as well as when ran in CI. Right now it's branching based off the previous commit, if it was a merge commit then it's assumed it's CI, but all other commits are viewed as local changes.

We should change this logic, it's not clear and I think it's flakey.

Reply:
I came up with this solution to be the simplest one:
```python
repository = Repository(CWD)
head = repository.head.peel(
return repository.diff(head.parents[0], head, context_lines=0)
```